File: claude-explore/tools/iff.py
"""
Strict IFF-85 reader for Design-It! 3-D VVR / WLB / TLB files.

Correctness oracle: every container's payload must be consumed EXACTLY by its
children, with zero leftover bytes. If the schema below is wrong, parsing fails
loudly instead of silently guessing.
"""
import struct, os

# Payload is 4-byte form-type followed by chunks.
FORMLIKE = {'FORM', 'CAT ', 'LIST'}

# Payload is pure chunks, no header bytes.
CONTAINERS = {'ROOT', 'PRSM', 'PGRP', 'PREF', 'SUTX', 'VGER'}

# Payload is N header bytes, then chunks.
CONTAINERS_HDR = {'SURF': 2}

# (parent, tag) -> header byte count. A FEAT inside a SURF carries a 2-byte
# header (the side selector); a FEAT that IS a 2D library clip does not.
CONTAINERS_HDR_CTX = {('SURF', 'FEAT'): 2}


class Chunk:
    __slots__ = ('tag', 'formtype', 'subtype', 'data', 'children', 'hdr',
                 'offset', 'trailing')

    def __init__(self, tag, data, offset=0):
        self.tag = tag
        self.formtype = None
        self.subtype = None
        self.data = data        # leaf payload (or header bytes for CONTAINERS_HDR)
        self.children = []
        self.hdr = b''
        self.offset = offset
        self.trailing = 0   # bytes at the end of this container we could not parse

    def kids(self, tag):
        return [c for c in self.children if c.tag == tag]

    def kid(self, tag):
        for c in self.children:
            if c.tag == tag:
                return c
        return None

    def find_all(self, tag):
        out = []
        if self.tag == tag:
            out.append(self)
        for c in self.children:
            out.extend(c.find_all(tag))
        return out

    def __repr__(self):
        t = f'{self.tag}<{self.formtype}>' if self.formtype else self.tag
        if self.subtype:
            t += f':{self.subtype}'
        return f'<{t} data={len(self.data)} kids={len(self.children)}>'


class IFFError(Exception):
    pass


def tag_bad(tag):
    return not all(('A' <= ch <= 'Z') or ch == ' ' for ch in tag)


def _hdrlen(tag, parent_tag):
    if (parent_tag, tag) in CONTAINERS_HDR_CTX:
        return CONTAINERS_HDR_CTX[(parent_tag, tag)]
    return CONTAINERS_HDR.get(tag)


def _parse_seq(buf, start, end, parent, strict=True):
    """Parse a sequence of chunks in buf[start:end] into parent.children."""
    off = start
    while off < end:
        if off + 8 > end:
            if strict:
                raise IFFError(f'truncated chunk header at {off} (end {end})')
            return
        tag = buf[off:off + 4].decode('latin1')
        ln = struct.unpack('>I', buf[off + 4:off + 8])[0]
        body = off + 8
        stop = body + ln
        if tag_bad(tag) or stop > end:
            # Known real-world defect: a handful of shipped ID*.WLB gallery
            # items have short/stale length fields, leaving stray bytes behind.
            # In tolerant mode, abandon the rest of this container rather than
            # inventing structure.
            if not strict:
                parent.trailing = end - off
                return
            raise IFFError(f'{tag!r} at {off} declares {ln} bytes, overruns end {end}')
        c = Chunk(tag, b'', off)
        if ln == 0:
            # Zero-length chunks are type markers (e.g. FEAT/PRSM inside a WLB
            # VCLP item declare what kind of library entry it is).
            pass
        elif tag in FORMLIKE:
            c.formtype = buf[body:body + 4].decode('latin1')
            p2 = body + 4
            if c.formtype == 'VCLP':
                # A library clip carries a bare 4-byte subtype token (FEAT for a
                # 2D gallery item, PRSM for a 3D one) right after the form type,
                # with no length field of its own.
                c.subtype = buf[p2:p2 + 4].decode('latin1')
                p2 += 4
            _parse_seq(buf, p2, stop, c, strict)
        elif tag in CONTAINERS:
            _parse_seq(buf, body, stop, c, strict)
        elif _hdrlen(tag, parent.tag) is not None:
            n = _hdrlen(tag, parent.tag)
            c.hdr = buf[body:body + n]
            _parse_seq(buf, body + n, stop, c, strict)
        elif tag in CONTAINERS or tag == 'FEAT':
            _parse_seq(buf, body, stop, c, strict)
        else:
            c.data = buf[body:stop]
        parent.children.append(c)
        off = stop


def parse(buf):
    """Parse a whole file buffer. Returns a synthetic root Chunk."""
    root = Chunk('$ROOT', b'')
    # Top level: walk to EOF, ignoring any declared top-chunk length that lies.
    off = 0
    n = len(buf)
    while off + 8 <= n:
        tag = buf[off:off + 4].decode('latin1')
        ln = struct.unpack('>I', buf[off + 4:off + 8])[0]
        stop = min(off + 8 + ln, n)
        c = Chunk(tag, b'', off)
        if tag in FORMLIKE:
            c.formtype = buf[off + 8:off + 12].decode('latin1')
            if tag == 'CAT ':
                # CAT lengths are unreliable in the shipped gallery files, and
                # a few individual items under-report their own length too.
                # Locate items by signature and bound each by the next one.
                stop = n
                starts = []
                i = off + 12
                while True:
                    i = buf.find(b'FORM', i)
                    if i < 0 or i >= n:
                        break
                    if buf[i + 8:i + 12] == b'VCLP':
                        starts.append(i)
                        i += 12
                    else:
                        i += 4
                for si, s0 in enumerate(starts):
                    s1 = starts[si + 1] if si + 1 < len(starts) else n
                    ln2 = struct.unpack('>I', buf[s0 + 4:s0 + 8])[0]
                    _parse_seq(buf, s0, min(s0 + 8 + ln2, s1), c, strict=False)
                    if c.children:
                        # re-bound the item we just added to the signature gap
                        pass
            else:
                _parse_seq(buf, off + 12, stop, c)
        elif tag in CONTAINERS:
            _parse_seq(buf, off + 8, stop, c)
        else:
            c.data = buf[off + 8:stop]
        root.children.append(c)
        off = stop
    return root


def load(path):
    with open(path, 'rb') as f:
        return parse(f.read())


# ---------- typed readers ----------

def i32(b, o):
    return struct.unpack_from('>i', b, o)[0]

def u32(b, o):
    return struct.unpack_from('>I', b, o)[0]

def i16(b, o):
    return struct.unpack_from('>h', b, o)[0]

def u16(b, o):
    return struct.unpack_from('>H', b, o)[0]

def fp(b, o):
    """16.16 fixed point."""
    return struct.unpack_from('>i', b, o)[0] / 65536.0

def f64(b, o):
    return struct.unpack_from('>d', b, o)[0]

def pstring(b, o):
    n = b[o]
    return b[o + 1:o + 1 + n].rstrip(b'\x00').decode('latin1'), o + 1 + n


def walk_files(rootdir, exts=('.VVR', '.WLB', '.TLB')):
    for dirpath, _, files in os.walk(rootdir):
        if '.git' in dirpath:
            continue
        for f in sorted(files):
            if f.upper().endswith(exts):
                yield os.path.join(dirpath, f)

File: web/src/iff.js
/**
 * Strict EA IFF-85 reader for Design-It! 3-D .VVR / .WLB / .TLB files.
 *
 * Correctness oracle: every container's payload must be consumed exactly by
 * its children. A wrong schema fails loudly instead of guessing.
 */

const FORMLIKE = new Set(['FORM', 'CAT ', 'LIST']);
const CONTAINERS = new Set(['ROOT', 'PRSM', 'PGRP', 'PREF', 'SUTX', 'VGER']);
const CONTAINERS_HDR = { SURF: 2 };
// A FEAT inside a SURF carries a 2-byte header (the side selector); a FEAT
// that IS a 2D library clip does not.
const CONTAINERS_HDR_CTX = { 'SURF/FEAT': 2 };

export class Chunk {
  constructor(tag, offset = 0) {
    this.tag = tag;
    this.formtype = null;
    this.subtype = null;
    this.data = null;      // DataView over the payload, for leaves
    this.hdr = null;       // DataView over header bytes, for SURF/FEAT
    this.children = [];
    this.offset = offset;
  }
  kid(tag) { return this.children.find((c) => c.tag === tag) || null; }
  kids(tag) { return this.children.filter((c) => c.tag === tag); }
  findAll(tag, out = []) {
    if (this.tag === tag) out.push(this);
    for (const c of this.children) c.findAll(tag, out);
    return out;
  }
}

export class IFFError extends Error {}

const dec = new TextDecoder('latin1');
const tagAt = (buf, o) => dec.decode(new Uint8Array(buf, o, 4));

function hdrLen(tag, parentTag) {
  const ctx = CONTAINERS_HDR_CTX[`${parentTag}/${tag}`];
  if (ctx !== undefined) return ctx;
  return CONTAINERS_HDR[tag];
}

function parseSeq(buf, view, start, end, parent, strict) {
  let off = start;
  while (off < end) {
    if (off + 8 > end) {
      if (strict) throw new IFFError(`truncated chunk header at ${off}`);
      return;
    }
    const tag = tagAt(buf, off);
    const len = view.getUint32(off + 4, false);
    const body = off + 8;
    let stop = body + len;
    if (stop > end) {
      // A few shipped ID*.WLB clips under-report their own length.
      if (!strict) { parent.trailing = end - off; return; }
      throw new IFFError(`${tag} at ${off} declares ${len} bytes, overruns ${end}`);
    }
    const c = new Chunk(tag, off);
    const hl = hdrLen(tag, parent.tag);
    if (len === 0) {
      // zero-length chunks are type markers, not containers
    } else if (FORMLIKE.has(tag)) {
      c.formtype = tagAt(buf, body);
      let p2 = body + 4;
      if (c.formtype === 'VCLP') { c.subtype = tagAt(buf, p2); p2 += 4; }
      parseSeq(buf, view, p2, stop, c, strict);
    } else if (hl !== undefined) {
      c.hdr = new DataView(buf, body, hl);
      parseSeq(buf, view, body + hl, stop, c, strict);
    } else if (CONTAINERS.has(tag) || tag === 'FEAT') {
      parseSeq(buf, view, body, stop, c, strict);
    } else {
      c.data = new DataView(buf, body, stop - body);
    }
    parent.children.push(c);
    off = stop;
    // IFF-85 pads an odd-length chunk to an even boundary, and this format does
    // too -- it simply never came up. Every chunk in the SoftKey corpus happens
    // to have an even length, so "exact lengths, no padding" held for 393 files
    // and got written into the spec as fact. The Virtus VRML content breaks it:
    // VRAN carries a URL, and `file:///kitchen.wrl` is 19 bytes followed by a
    // 00 before the next chunk. Without this, three tutorial files fail to parse.
    if (len % 2 && off < end && view.getUint8(off) === 0) off += 1;
  }
}

/** Parse a whole file. Returns a synthetic root chunk. */
export function parse(buf) {
  const view = new DataView(buf);
  const root = new Chunk('$ROOT');
  const n = buf.byteLength;
  let off = 0;
  while (off + 8 <= n) {
    const tag = tagAt(buf, off);
    const len = view.getUint32(off + 4, false);
    let stop = Math.min(off + 8 + len, n);
    const c = new Chunk(tag, off);
    if (FORMLIKE.has(tag)) {
      c.formtype = tagAt(buf, off + 8);
      if (tag === 'CAT ') {
        // CAT lengths are stale in the shipped gallery files. Find clips by
        // signature and bound each by the start of the next.
        stop = n;
        const starts = [];
        let i = off + 12;
        const bytes = new Uint8Array(buf);
        while (i < n - 12) {
          if (bytes[i] === 0x46 && bytes[i + 1] === 0x4f && bytes[i + 2] === 0x52 && bytes[i + 3] === 0x4d &&
              bytes[i + 8] === 0x56 && bytes[i + 9] === 0x43 && bytes[i + 10] === 0x4c && bytes[i + 11] === 0x50) {
            starts.push(i); i += 12;
          } else i += 1;
        }
        for (let si = 0; si < starts.length; si++) {
          const s0 = starts[si];
          const s1 = si + 1 < starts.length ? starts[si + 1] : n;
          const l2 = view.getUint32(s0 + 4, false);
          parseSeq(buf, view, s0, Math.min(s0 + 8 + l2, s1), c, false);
        }
      } else {
        parseSeq(buf, view, off + 12, stop, c, true);
      }
    } else if (CONTAINERS.has(tag)) {
      parseSeq(buf, view, off + 8, stop, c, true);
    } else {
      c.data = new DataView(buf, off + 8, stop - (off + 8));
    }
    root.children.push(c);
    off = stop;
    if (len % 2 && off < n && view.getUint8(off) === 0) off += 1;   // odd-length pad
  }
  return root;
}

// ---- typed readers ----
export const i32 = (d, o) => d.getInt32(o, false);
export const u32 = (d, o) => d.getUint32(o, false);
export const i16 = (d, o) => d.getInt16(o, false);
export const u16 = (d, o) => d.getUint16(o, false);
/** 16.16 fixed point -- the format's universal numeric type. */
export const fp = (d, o) => d.getInt32(o, false) / 65536;
export const f64 = (d, o) => d.getFloat64(o, false);

export function pstring(d, o = 0) {
  const n = d.getUint8(o);
  let s = '';
  for (let i = 0; i < n; i++) {
    const c = d.getUint8(o + 1 + i);
    if (c) s += String.fromCharCode(c);
  }
  return s;
}

/** Named clips inside a .WLB gallery. */
export function wlbItems(root) {
  const out = [];
  for (const cat of root.children) {
    for (const it of cat.children) {
      if (it.tag !== 'FORM' || it.formtype !== 'VCLP') continue;
      const nm = it.kid('NAME');
      // `root` rides along: the gallery's TXTB sits above the clips, so a
      // clip built on its own still needs a way back to the texture table.
      out.push({ name: nm ? pstring(nm.data) : '?', kind: it.subtype, chunk: it, root });
    }
  }
  return out;
}
